### PR TITLE
First draft for state retention

### PR DESCRIPTION
**Pull request by TCke83**
New pull request with same changes + added tests.

The LoxoneState class needs some help for testing, i'm still a newbie with the used mocking framework and with kotlin code.

### CODE REVIEW

**Comment by jimirocks** [comment]
Hi, that's a lot of code. It will take a while to review. But well, thanks for contribution.

**Comment by TCke83** [comment]
It's the same code change as PR #174, but it took me a while to find the time to write the tests

File: src/main/java/cz/smarteon/loxone/app/Control.java [hunk 1]
```diff
@@ -98,4 +100,18 @@ protected LoxoneUuids getCompulsoryState(final String stateName) {
         }
     }
 
+    /**
+     * Helper to get the jLocked state UUID for this control.
+     * If the control doesn't support locking, it will return null.
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/Control.java:105]: Consider adding `isLockable` or `supportsLocking` method for convenience (better checking boolean then nullabilty)

File: src/main/java/cz/smarteon/loxone/app/state/ControlState.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    void accept(@NotNull TextEvent event) {
+        // default implementation
+    }
+
+    /**
+     * Method to update the state of the command in the miniserver.
+     * @param controlCommand that needs to be sent
+     */
+    protected void sendCommand(ControlCommand<?> controlCommand) {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/ControlState.java:47]: his method seems redundant, please use `Loxone.sendControlCommand`  instead, see https://github.com/Smarteon/loxone-java/blob/master/src/main/java/cz/smarteon/loxone/Loxone.java#L261

File: src/main/java/cz/smarteon/loxone/app/state/DigitalInfoControlState.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    private void processActiveEvent(ValueEvent event) {
+        double value = event.getValue();
+        if (value == 0) {
+            state = false;
+            return;
+        } else if (value == 1) {
+            state = true;
+            return;
+        }
+        throw new IllegalArgumentException("Value " + value + " not known for DigitalInfoControl");
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/DigitalInfoControlState.java:52]: Just to consider - is there a real chance to reach this point? Could it be simplified to `return value == 1`?

File: src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    private Map<LoxoneUuid, ControlState<?>> controlStates;
+
+    static {
+        SUPPORTED_CONTROLS_STATE_MAP = new HashMap<>();
+        SUPPORTED_CONTROLS_STATE_MAP.put(SwitchControl.class, SwitchControlState.class);
+        SUPPORTED_CONTROLS_STATE_MAP.put(AnalogInfoControl.class, AnalogInfoControlState.class);
+        SUPPORTED_CONTROLS_STATE_MAP.put(DigitalInfoControl.class, DigitalInfoControlState.class);
+    }
+
+    public LoxoneState(LoxoneProfile loxoneProfile) {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java:45]: well, consider having another constructor taking `Loxone` as parameter - it will ease the testing and also allow user to use single instance of `Loxone` class for both state and other usecases.  

You can keep this constructor for convenience.

Mark parameter nullability and check if needed.

File: src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        loxone = new Loxone(loxoneProfile);
+        loxone.setEventsEnabled(true);
+        loxone.registerLoxoneAppListener(this);
+
+    }
+
+    public Loxone loxone() {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java:52]: mark nulability

File: src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        return loxone;
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> T getStateForControl(Control control) {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/LoxoneState.java:57]: mark nullability

File: src/main/java/cz/smarteon/loxone/app/state/SwitchControlState.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            processActiveEvent(event);
+        }
+    }
+
+    /**
+     * Process the ValueEvent as an active state event message and update the state of the control accordingly.
+     * @param event value event received
+     */
+    private void processActiveEvent(ValueEvent event) {
+        double value = event.getValue();
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/SwitchControlState.java:70]: consider intepreting anything besides `1` as false.

File: src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Value;
+import lombok.extern.jackson.Jacksonized;
+
+@Value
+@Builder
+@Jacksonized
+public class LockedEvent {
+
+    @Builder.Default
+    Locked locked = Locked.NO;
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java:14]: not sure how the lombok works here, but I guess it won't make the fields private right? So let's make the private please

File: src/main/java/cz/smarteon/loxone/message/LoxoneEvent.java [hunk 9]
```diff
@@ -1,15 +1,23 @@
...
+import lombok.Getter;
+import lombok.ToString;
 import org.jetbrains.annotations.NotNull;
 
 import static java.util.Objects.requireNonNull;
 
 /**
  * Common predecessor of loxone events.
  */
+@Getter
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/message/LoxoneEvent.java:13]: pls move this lombok refactorings to separate commit

File: src/main/java/cz/smarteon/loxone/message/TextEvent.java [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
+import lombok.ToString;
 import org.jetbrains.annotations.NotNull;
 
 import static java.util.Objects.requireNonNull;
 
 /**
  * Loxone event carrying text value.
  */
+@Getter
+@ToString(callSuper = true)
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/message/TextEvent.java:14]: pls move this lombok refactorings to separate commit

File: src/main/java/cz/smarteon/loxone/message/ValueEvent.java [hunk 11]
```diff
@@ -1,15 +1,22 @@
...
 import cz.smarteon.loxone.LoxoneUuid;
+import lombok.Getter;
+import lombok.ToString;
 import org.jetbrains.annotations.NotNull;
 
 /**
  * Loxone event carrying numeric value.
  */
+@Getter
+@ToString(callSuper = true)
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/message/ValueEvent.java:12]: pls move this lombok refactorings to separate commit

File: src/test/kotlin/app/state/AnalogInfoControlStateTest.kt [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Test
+    fun `should process value event`() {
+        val loxone = mockk<Loxone> {}
+        val analogInfoControl = mockk<AnalogInfoControl> {
+            every { stateValue() } returns LoxoneUuid("0f869a64-028d-0cc2-ffffd4c75dbaf53c")
+        }
+        val analogInfoControlState = AnalogInfoControlState(loxone, analogInfoControl);
+
+        expectThat(analogInfoControlState.value).equals(null)
```
Comment (jimirocks) [inline src/test/kotlin/app/state/AnalogInfoControlStateTest.kt:28]: consider using ParneterizedTest

**Review by jimirocks — CHANGES_REQUESTED** [review]
Hi, viewing the code second time brought few more ideas - pls see the comments. It's nothing serious or complex, so should be easy to resolve.

Regarding tests, my colleague @knotekt  will try to help you with some code hints soon.

**Comment by jimirocks** [comment]
So we put some guidance for testing here: https://github.com/TCke83/loxone-java/pull/1/files  please take a look, we can discuss further...

File: src/main/java/cz/smarteon/loxone/app/state/DigitalInfoControlState.java (hunk 3 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/app/state/DigitalInfoControlState.java:52]: The ValueEvent gets a double, could be any value in theory, but I suppose in real-life it will always be 0 or 1 for a DigitalInfoControlState, could simplify it.

File: src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java (hunk 8 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java:14]: Lombok makes them private, so private modifier is not needed

**Comment by TCke83** [comment]
Added a few tests for LoxoneState class

**Comment by jimirocks** [comment]
Hi @TCke83 do you plan to work more  on this yet, or should we do final walkthrough? In that case pls. just reorganize the commits - one commit with lombok refactorings and everything else could be in the second one.

**Comment by TCke83** [comment]
I've added a test for the constructor check. The exception flow in the initializeState() method is covered in test 'controlStates should be compatible'. I don't really see a way to test the initializeState() method further, if you think this should be done, some help would be welcome.

**Comment by jimirocks** [comment]
Please take a look at https://github.com/TCke83/loxone-java/pull/2  I finished the changes my colleague proposed as guidance. Please take a look and take notes for the future. The exceptions in LoxoneState are still not covered, but it's probably OK for now - can be added in future development if needed. Once you accept my changes, squash it into single commit and I am ready to merge and release.

**Comment by TCke83** [comment]
I've looked at the test and merged & squashed them in the pull request. Thanks for the help and guidance.

**Review by jimirocks — APPROVED** [review]
